icu4c/source/common/uloc.cpp

```diff
@@ -1485,7 +1490,12 @@ _canonicalize(const char* localeID,
     }
 
     if (_hasBCP47Extension(localeID)) {
-        _ConvertBCP47(tmpLocaleID, localeID, tempBuffer, sizeof(tempBuffer), err);
+        int32_t convertedSize = _ConvertBCP47(tmpLocaleID, localeID, tempBuffer, sizeof(tempBuffer), err);
+        if (*err == U_BUFFER_OVERFLOW_ERROR || *err == U_STRING_NOT_TERMINATED_WARNING) {
+            heapTempBuffer.adoptInstead((char*)uprv_malloc(convertedSize + 2));
```

Comment:
I think we also need to check if `uprv_malloc` fails here, as `_ConvertBCP47` won't set `err` to `U_MEMORY_ALLOCATION_ERROR` if the buffer is NULL/nullptr.

Reply:
agree with jeff

Reply:
Done.  See comments above.

Reply:
nit: You added a `static_cast<char*>` in the other file, but not here. ;)